### PR TITLE
fix: race condition that creates corrupted scripts

### PR DESCRIPTION
**Pull request by KSXGitHub**
Fixes https://github.com/pnpm/pnpm/issues/7833

### CODE REVIEW

File: pkg-manager/link-bins/src/index.ts [hunk 1]
```diff
@@ -166,6 +152,35 @@ async function _linkBins (
...
+function deduplicateCommands (commands: CommandInfo[], binsDir: string): CommandInfo[] {
+  const cmdGroups = groupBy(cmd => cmd.name, commands)
+  const result: CommandInfo[] = []
+  for (const group of Object.values(cmdGroups)) {
+    if (!group || group.length === 0) continue
+    if (group.length === 1) {
+      result.push(group[0])
+      continue
+    }
+    const withOwnName = group.find(cmd => cmd.ownName) // prefer bin with own name if exist
```
Comment (gluxon) [inline pkg-manager/link-bins/src/index.ts]: The decision to prefer bins with their own name is clever. I like it.

Comment (KSXGitHub) [reply pkg-manager/link-bins/src/index.ts]: It wasn't my decision. I just fixed failed tests.

File: pkg-manager/link-bins/src/index.ts [hunk 2]
```diff
@@ -166,6 +157,43 @@ async function _linkBins (
...
+    return chosen
+  })
+}
+
+function compareCommandsInConflict (a: CommandInfo, b: CommandInfo): -1 | 0 | 1 {
+  if (a.ownAlias && b.ownAlias === false) return 1
+  if (a.ownAlias === false && b.ownAlias) return -1
+  if (a.ownName && !b.ownName) return 1
+  if (!a.ownName && b.ownName) return -1
+  if (a.pkgName !== b.pkgName) return 0 // it's pointless to compare versions of 2 different package
```
Comment (zkochan) [inline pkg-manager/link-bins/src/index.ts]:
```suggestion
  if (a.pkgName !== b.pkgName) return a.pkgName.localeCompare(b.pkgName) // it's pointless to compare versions of 2 different package
```

File: pkg-manager/link-bins/src/index.ts [hunk 3]
```diff
@@ -72,15 +73,16 @@ export async function linkBinsOfPkgsByAliases (
...
           isDirectDependency: directDependencies?.has(alias),
           nodeExecPath: opts.nodeExecPathByAlias?.[alias],
         }))
         .filter(({ depDir }) => !isSubdir(depDir, binsDir)) // Don't link own bins
-        .map(async ({ depDir, isDirectDependency, nodeExecPath }) => {
+        .map(async ({ pkgAlias, depDir, isDirectDependency, nodeExecPath }) => {
           const target = normalizePath(depDir)
           const cmds = await getPackageBins(pkgBinOpts, target, nodeExecPath)
-          return cmds.map((cmd) => ({ ...cmd, isDirectDependency }))
+          return cmds.map((cmd) => ({ ...cmd, isDirectDependency, ownAlias: cmd.name === pkgAlias, pkgAlias }))
```
Comment (zkochan) [inline pkg-manager/link-bins/src/index.ts]: I am not sure we need to consider aliases.

Comment (KSXGitHub) [reply pkg-manager/link-bins/src/index.ts]: I have removed alias from consideration

**Comment by zkochan** [comment]
It is probably all good now. I'll review it later.

**Review by zkochan — APPROVED** [review]

**Pull request merged**